llvm/lib/Target/AIE/AIEMachineScheduler.cpp

```diff
@@ -479,6 +483,56 @@ int AIEPostRASchedStrategy::getMaxDeltaCycles(const SchedBoundary &Zone) const {
                    BottomUpDelta.getValue()});
 }
 
+bool AIEPostRASchedStrategy::canOptimizeMemoryAccess(SUnit &SU,
+                                                     SchedBoundary &Zone,
+                                                     const int DeltaCycle) {
+  if (!IgnoreMemoryBankConflict)
+    return false;
+
+  if (!SU.getInstr()->mayLoadOrStore()) {
+    return false;
+  }
+
+  const int MinDelta = -getMaxDeltaCycles(Zone);
+  if (!(DeltaCycle - 1 >= MinDelta))
+    return false;
+
+  const AIEBaseMCFormats &Formats = *getTII(*Zone.DAG)->getFormatInterface();
+  AIEHazardRecognizer &HR = *getAIEHazardRecognizer(Zone);
+  MachineInstr *MI = SU.getInstr();
+
+  const std::vector<unsigned int> *AlternateOpcodes;
+  auto DefaultOpcode = std::vector<unsigned int>{SU.getInstr()->getOpcode()};
+  AlternateOpcodes =
+      Formats.getAlternateInstsOpcode(SU.getInstr()->getOpcode())
+          ? Formats.getAlternateInstsOpcode(SU.getInstr()->getOpcode())
+          : &DefaultOpcode;
+
+  unsigned int OpcodeWithMemoryBankConflict = 0;
+  for (const unsigned int AltOpcode : *AlternateOpcodes) {
+    // Check if the conflict was caused by a memory bank.
+    if (HR.checkConflict(*MI, getTII(*Zone.DAG)->get(AltOpcode), DeltaCycle) ==
+        AIEHazardRecognizer::ConflictType::MemoryBank) {
+      OpcodeWithMemoryBankConflict = AltOpcode;
+      break;
+    }
+  }
+  // Check if the memory operation will also have a conflict in the next cycle.
```

Comment:
check: next cycle or prev cycle? I always get confused with bottom-up scheduling.

Reply:
ha ha same happens with me, I wrote "next cycle" in perspective with bottom up which is previous cycles for top-down.